server/src/main/java/org/elasticsearch/persistent/AllocatedPersistentTask.java

```diff
@@ -57,6 +57,20 @@ public void updatePersistentTaskState(final PersistentTaskState state,
         persistentTasksService.sendUpdateStateRequest(persistentTaskId, allocationId, state, listener);
     }
 
+    /**
+     * Unassign the corresponding persistent task from its current node.
+     * <p>
+     * This will trigger an attempt to reassign the task. If there is nothing
+     * in the task assignment logic to prevent it being assigned back to the
+     * same node then this may happen. It is expected that callers will have
+     * done something that will prevent the task assignment logic immediately
+     * reassigning the task back to the same node.
+     */
+    public void unassign(final String reason,
```

Comment:
Can we instead have a method named `markAsAbortedLocally` that both unregisters the task and notifies the master that the task was locally aborted (sort of the same as unassigned)? It completes the task's local lifecycle just like calling `markAsCompleted` and `markAsFailed`.

That removes the need for a listener here, since the task should no longer do anything after this call. And it ensures:

1. Clearer semantics - this can only be called to terminate the tasks lifecycle. Without this, the task would still need to complete itself to unregister from the task manager.
2. There is no race between unassign and task completion. I can imagine some doing unassign and then just completing the task, which could lead to the task completing rather than unassigning if the messages are delivered out of order.